File: skills/clean-ddd-hexagonal/examples/typescript/infrastructure/in_memory_order_repository.ts
// infrastructure/adapters/driven/in_memory/order_repository.ts
import { IOrderRepositoryPort } from '@/application/ports/driven/order_repository_port';
import { Order } from '@/domain/order/order';
import { OrderId } from '@/domain/order/value_objects';

/**
 * In-memory adapter for Order persistence, primarily for testing.
 *
 * This implementation stores orders in a Map, providing fast operations
 * without external dependencies. It's ideal for unit tests where database
 * isolation is desired.
 *
 * @example
 * ```typescript
 * const repo = new InMemoryOrderRepository();
 * await repo.save(order);
 *
 * // In tests
 * expect(repo.getAll()).toHaveLength(1);
 * repo.clear();
 * ```
 */
export class InMemoryOrderRepository implements IOrderRepositoryPort {
  /** Internal storage mapping order IDs to Order objects. */
  private orders: Map<string, Order> = new Map();

  /**
   * Find an Order by its identifier.
   *
   * @param id - The OrderId to search for
   * @returns The Order if found, null otherwise
   */
  async findById(id: OrderId): Promise<Order | null> {
    return this.orders.get(id.value) ?? null;
  }

  /**
   * Store an Order in memory.
   *
   * @param order - The Order to persist
   */
  async save(order: Order): Promise<void> {
    this.orders.set(order.id.value, order);
  }

  /**
   * Remove an Order from memory.
   *
   * @param order - The Order to delete
   */
  async delete(order: Order): Promise<void> {
    this.orders.delete(order.id.value);
  }

  // Test helpers

  /**
   * Remove all orders from the repository.
   *
   * Test helper method not part of the repository interface.
   */
  clear(): void {
    this.orders.clear();
  }

  /**
   * Retrieve all stored orders.
   *
   * Test helper method not part of the repository interface.
   *
   * @returns Array of all orders in the repository
   */
  getAll(): Order[] {
    return Array.from(this.orders.values());
  }
}

File: skills/clean-ddd-hexagonal/examples/python/infrastructure/in_memory_order_repository.py
# src/infrastructure/in_memory/order_repository.py
"""In-memory implementation of the Order repository for testing.

This module provides a simple in-memory adapter for the OrderRepository
interface, suitable for unit tests and local development without a database.
"""
from typing import Optional, Dict, List
from domain.order.aggregate import Order
from domain.order.repository import OrderRepository
from domain.order.value_objects import OrderId


class InMemoryOrderRepository(OrderRepository):
    """In-memory adapter for Order persistence, primarily for testing.

    This implementation stores orders in a dictionary, providing fast
    operations without external dependencies. It's ideal for unit tests
    where database isolation is desired.

    Attributes:
        _orders: Internal storage mapping order ID strings to Order objects.
    """

    def __init__(self) -> None:
        """Initialize an empty in-memory repository."""
        self._orders: Dict[str, Order] = {}

    async def find_by_id(self, id: OrderId) -> Optional[Order]:
        """Retrieve an Order by its identifier.

        Args:
            id: The OrderId to search for.

        Returns:
            The Order if found, None otherwise.
        """
        return self._orders.get(id.value)

    async def save(self, order: Order) -> None:
        """Store an Order in memory.

        Args:
            order: The Order to persist.
        """
        self._orders[order.id.value] = order

    async def delete(self, order: Order) -> None:
        """Remove an Order from memory.

        Args:
            order: The Order to delete.
        """
        self._orders.pop(order.id.value, None)

    # Test helpers

    def clear(self) -> None:
        """Remove all orders from the repository.

        This is a test helper method not part of the repository interface.
        """
        self._orders.clear()

    def get_all(self) -> List[Order]:
        """Retrieve all stored orders.

        This is a test helper method not part of the repository interface.

        Returns:
            List of all orders currently in the repository.
        """
        return list(self._orders.values())
